packages/protocol/contracts/L1/libs/LibProving.sol

```diff
@@ -6,371 +6,228 @@
+            bytes32 instance = _getInstance(
+                evidence,
+                resolver.resolve("signal_service", false),
+                resolver.resolve(config.chainId, "signal_service", false)
+            );
+            bool verified = _verifyZKProof(
+                resolver,
+                evidence.zkproof,
+                instance
+            );
+            if (!verified) revert L1_BLOCK_PROOF();
+        }
     }
 
     function proveBlockInvalid(
         TaikoData.State storage state,
         TaikoData.Config memory config,
         AddressResolver resolver,
         uint256 blockId,
-        bytes[] calldata inputs
-    ) public {
-        // Check and decode inputs
-        if (inputs.length != 3) revert L1_INPUT_SIZE();
-        TaikoData.Evidence memory evidence = abi.decode(
-            inputs[0],
-            (TaikoData.Evidence)
-        );
-        TaikoData.BlockMetadata memory target = abi.decode(
-            inputs[1],
-            (TaikoData.BlockMetadata)
+        bytes calldata evidenceBytes
+    ) internal {
+        TaikoData.InvalidBlockEvidence memory evidence = abi.decode(
+            evidenceBytes,
+            (TaikoData.InvalidBlockEvidence)
         );
 
-        // Check evidence
-        if (evidence.meta.id != blockId) revert L1_ID();
-        if (evidence.proofs.length != 2) revert L1_PROOF_LENGTH();
+        TaikoData.BlockMetadata memory meta = evidence.meta;
+        _checkMetadata(state, config, meta, blockId);
 
-        IProofVerifier proofVerifier = IProofVerifier(
-            resolver.resolve("proof_verifier", false)
-        );
-
-        if (config.enableAnchorValidation) {
-            _proveAnchorForInvalidBlock({
-                config: config,
-                resolver: resolver,
-                target: target,
-                proofVerifier: proofVerifier,
-                evidence: evidence,
-                invalidateBlockReceipt: inputs[2]
-            });
-        }
+        if (
+            LibUtils.hashZKProof(abi.encode(evidence.zkproof)) !=
+            meta.txListProofHash
+        ) revert L1_TX_LIST_PROOF();
 
-        // ZK-prove block and mark block proven as invalid.
-        _proveBlock({
+        bool oracleProving = _proveBlock({
             state: state,
-            config: config,
             resolver: resolver,
-            proofVerifier: proofVerifier,
-            evidence: evidence,
-            target: target,
-            blockHashOverride: LibUtils.BLOCK_DEADEND_HASH
+            blockId: blockId,
+            parentHash: evidence.parentHash,
+            snippet: Snippet(LibUtils.BLOCK_DEADEND_HASH, 0),
+            prover: msg.sender
         });
+
+        if (!oracleProving && !config.skipZKPVerification) {
+            bool verified = _verifyZKProof(
+                resolver,
+                evidence.zkproof,
+                meta.txListHash
+            );
+            if (verified) revert L1_TX_LIST_PROOF();
```

Comment:
Since in `_verifyZKProof`, we have:
```solidity
// Do not revert when circuitId is invalid.
address verifier = resolver.resolve(
    string.concat("verifier_", Strings.toString(zkproof.circuitId)),
    true
);
if (verifier == address(0)) return false;
```
so if a prover can pass the check here by giving an invalid `circuitId`?

Reply:
True. circuit ID is considered as part of the ZKP. the `txListProof` is hashed into a `txListProofHash`, then `txListProofHash` is hashed into `instance`. So if the circuit ID is wrong, the main ZK proof cannot be generated.